content/3_Silver/DFS.mdx

```diff
@@ -585,3 +585,77 @@ public class BuildingTeams
 ### Problems
 
 <Problems problems="bip" hideSuggestProblemButton />
+
+## Quiz
+<Quiz>
+	<Quiz.Question>
+		What would be a valid function header for DFS on an undirected tree to work in linear time *without* any other variables?
+		<Quiz.Answer correct>
+			`void dfs(int v, int p, vector<vector<int> >& adj)`
+			<Quiz.Explanation>
+				Correct! For a tree, we do not need a visited array. We only need to check if we are revisiting the parent node, so `v` as a current node and `p` as a parent node would work.
+
+				Calling `adj` as a `vector<vector<int> >&` instead of just `vector<vector<int> >` ensures that the pointer is copied over each time, which takes constant instead of linear time. Therefore, the entire DFS would take linear time.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+		<Quiz.Answer>
+			`void dfs(int v, int p, vector<vector<int> > adj)`
+			<Quiz.Explanation>
+				Incorrect. This would not work in linear time because `adj` would be copied in $O(n)$ time every time DFS is called. As DFS is called $n$ times, this runs in $O(n^2)$.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+		<Quiz.Answer>
+			`void dfs(int v, vector<vector<int> > adj)`
+			<Quiz.Explanation>
+				Incorrect. This DFS function cannot check for a parent node, and therefore would never terminate.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+		<Quiz.Answer>
+			`void dfs(vector<vector<int> > adj)`
+			<Quiz.Explanation>
+				Incorrect. This DFS function cannot track the current node, therefore it cannot even run.
```

Comment:
technically you could append to the adjacency list to keep track of the current state of the DFS? Probably better to use `const vector<vector<int>>& adj`